### PR TITLE
migration to Juju 3.1, run tests on Juju 2.9

### PR DESCRIPTION
**Pull request by delgod**
in this PR:
* upgrade all dependencies to the latest version.
* update Juju to 3.1.0.1
* continue to run all integration tests on 2.9.29

### CODE REVIEW

**Comment by dmitry-ratushnyy** [comment]
Maybe we should add some descriptions to the MR ? :smile:

File: poetry.lock
Comment (dmitry-ratushnyy) [inline poetry.lock:1]: Is this file required to be committed?

Comment (delgod) [reply poetry.lock:1]: Yes, I want to have reproducible builds (be able to rebuild the application five years later with the same libraries).

**Review by MiaAltieri — APPROVED** [review]
Thanks Mykola!

**Review by dmitry-ratushnyy — APPROVED** [review]

**Review by dmitry-ratushnyy — APPROVED** [review]
